dg.go

```diff
@@ -50,13 +50,17 @@ func (d *directedGraph[NodeType]) Mermaid() string {
 			if isErrorPath {
 				errorPath = append(errorPath, connection)
 			} else {
-				result = append(result, connection)
+				successPath = append(successPath, connection)
 			}
 		}
 	}
 
-	result = append(result, errorPath...)
+	slices.Sort(successPath)
+	slices.Sort(errorPath)
```

Comment:
My only concern is that this (presumably) does an alpha-numeric sort, which is an arbitrary way to sort the paths, and the order does seem to affect the details of how a particular mermaid renderer will display the flowchart. I don't have an alternative to propose, and this does have the advantage of consistency.

Reply:
Here is a rendered workflow with the current mermaid output:
```mermaid
%% Mermaid markdown workflow
flowchart LR
%% Success path
steps.metadata.outputs.success-->steps.example.starting
steps.metadata.cancelled-->steps.metadata.outputs
steps.metadata.starting-->steps.metadata.running
steps.metadata.starting-->steps.metadata.starting.started
steps.example.disabled-->steps.example.disabled.output
steps.example.deploy-->steps.example.starting
steps.metadata.running-->steps.metadata.outputs
steps.example.running-->steps.example.outputs
steps.metadata.outputs-->steps.metadata.outputs.success
steps.example.outputs-->steps.example.outputs.success
steps.metadata.enabling-->steps.metadata.disabled
steps.metadata.enabling-->steps.metadata.enabling.resolved
steps.metadata.enabling-->steps.metadata.starting
steps.example.outputs.success-->outputs.success
steps.metadata.deploy-->steps.metadata.starting
steps.example.starting-->steps.example.starting.started
steps.example.starting-->steps.example.running
steps.metadata.disabled-->steps.metadata.disabled.output
steps.example.cancelled-->steps.example.outputs
steps.example.enabling-->steps.example.enabling.resolved
steps.example.enabling-->steps.example.starting
steps.example.enabling-->steps.example.disabled
%% Error path
steps.metadata.cancelled-->steps.metadata.deploy_failed
steps.metadata.cancelled-->steps.metadata.crashed
steps.example.crashed-->steps.example.crashed.error
steps.metadata.starting-->steps.metadata.crashed
steps.example.deploy-->steps.example.deploy_failed
steps.metadata.crashed-->steps.metadata.crashed.error
steps.metadata.running-->steps.metadata.crashed
steps.example.deploy_failed-->steps.example.deploy_failed.error
steps.example.running-->steps.example.crashed
steps.metadata.outputs-->steps.metadata.outputs.error
steps.example.outputs-->steps.example.outputs.error
steps.metadata.enabling-->steps.metadata.crashed
steps.metadata.deploy_failed-->steps.metadata.deploy_failed.error
steps.metadata.deploy-->steps.metadata.deploy_failed
steps.example.starting-->steps.example.crashed
steps.example.cancelled-->steps.example.deploy_failed
steps.example.cancelled-->steps.example.crashed
steps.example.enabling-->steps.example.crashed
%% Mermaid end
```

Here is the same workflow with the mermaid sorted:
```mermaid
%% Mermaid markdown workflow
flowchart LR
%% Success path
steps.example.cancelled-->steps.example.outputs
steps.example.deploy-->steps.example.starting
steps.example.disabled-->steps.example.disabled.output
steps.example.enabling-->steps.example.disabled
steps.example.enabling-->steps.example.enabling.resolved
steps.example.enabling-->steps.example.starting
steps.example.outputs-->steps.example.outputs.success
steps.example.outputs.success-->outputs.success
steps.example.running-->steps.example.outputs
steps.example.starting-->steps.example.running
steps.example.starting-->steps.example.starting.started
steps.metadata.cancelled-->steps.metadata.outputs
steps.metadata.deploy-->steps.metadata.starting
steps.metadata.disabled-->steps.metadata.disabled.output
steps.metadata.enabling-->steps.metadata.disabled
steps.metadata.enabling-->steps.metadata.enabling.resolved
steps.metadata.enabling-->steps.metadata.starting
steps.metadata.outputs-->steps.metadata.outputs.success
steps.metadata.outputs.success-->steps.example.starting
steps.metadata.running-->steps.metadata.outputs
steps.metadata.starting-->steps.metadata.running
steps.metadata.starting-->steps.metadata.starting.started
%% Error path
steps.example.cancelled-->steps.example.crashed
steps.example.cancelled-->steps.example.deploy_failed
steps.example.crashed-->steps.example.crashed.error
steps.example.deploy_failed-->steps.example.deploy_failed.error
steps.example.deploy-->steps.example.deploy_failed
steps.example.enabling-->steps.example.crashed
steps.example.outputs-->steps.example.outputs.error
steps.example.running-->steps.example.crashed
steps.example.starting-->steps.example.crashed
steps.metadata.cancelled-->steps.metadata.crashed
steps.metadata.cancelled-->steps.metadata.deploy_failed
steps.metadata.crashed-->steps.metadata.crashed.error
steps.metadata.deploy_failed-->steps.metadata.deploy_failed.error
steps.metadata.deploy-->steps.metadata.deploy_failed
steps.metadata.enabling-->steps.metadata.crashed
steps.metadata.outputs-->steps.metadata.outputs.error
steps.metadata.running-->steps.metadata.crashed
steps.metadata.starting-->steps.metadata.crashed
%% Mermaid end
```

Reply:
My primary concern, obviously, is consistency, but a straightforward alphanumeric sort is not the only way to achieve that, and, since this is produced in a single, centralized place, we have the opportunity to implement anything we want...so, what do we want?

Looking at those two renderings, I concur that the second lacks somewhat aesthetically, but I'm not really sure why (perhaps just a cultural bias that things should flow to the right and _downward_); although, I do like the fact that the nodes which lack inbound edges are rendered with no nodes to their direct left (e.g., in the unsorted rendering, `steps.example.deploy` is surrounded, whereas it's a more clear origin point in the sorted rendering).

I don't see the `input` node...is there none (not even and "empty" one?)?  Or is it not shown because nothing depends on it?

Anyway, if we can't come up with an approach for "fixing" this, how about I merge it and we iterate later?

Reply:
The different visual spread of the two graphs is interesting, but I don't think either is necessarily preferable, and they're logical equivalent. And given that the order of the edges makes a difference to the graph generator, the "shapes" would seem to be completely coincidental anyway.

I'm not entirely sure that sorting the graph edges is going to help with markdown reviews anyway: edges can still move if nodes are renamed, even slightly. But at worst I don't think it hurts.

Reply:
> I don't think either is necessarily preferable

🤷  I was hoping that if we picked a particular node (to wit, `input`) and put that first, that it would result in a "nice shape" on a fairly reliable basis.  But, I was thwarted by the fact that the workflow that Dustin picked doesn't _have_ and `input` node!  😆

> the "shapes" would seem to be completely coincidental anyway.

Yeah, unless we can find something common (like `input`, but...) to anchor them.

> I'm not entirely sure that sorting the graph edges is going to help with markdown reviews anyway

It should help considerably in the (apparently typical 😉) case where the workflow is **not** modified.  In a case where the workflow _is_ modified, I'm hoping that it, combined with Git's desire to minimize displayed differences, will highlight only what actually changed.  (But, I could be dreamin'....  🫠)

> I don't think it hurts.

Same here.

Reply:
OK, based on [Jared's comment](https://github.com/arcalot/go-dgraph/pull/19#pullrequestreview-2154661639), I've enabled auto-merge, so, @dustinblack, when you resolve this conversation (hint, hint), this PR will merge.

Reply:
I don't actually have a strong opinion on which mermaid rendering is better, and I don't even think we can expect them to be deterministic or consistent if something with the renderer changes.  They both have the same information, and this code change at least partially helps with the consistency problem for code reviews, so I've nothing fundamentally against it.